flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -1650,6 +1679,54 @@ public void close() {
 		}
 	}
 
+	private static final class TransformingRocksIteratorWrapper extends RocksIteratorWrapper {
+		@Nonnull
+		private final StateSnapshotTransformer<byte[]> stateSnapshotTransformer;
+		private byte[] current;
+
+		public TransformingRocksIteratorWrapper(
+			@Nonnull RocksIterator iterator,
+			@Nonnull StateSnapshotTransformer<byte[]> stateSnapshotTransformer) {
+			super(iterator);
+			this.stateSnapshotTransformer = stateSnapshotTransformer;
+		}
+
+		@Override
+		public void seekToFirst() {
+			super.seekToFirst();
+			filterOrTransform(super::next);
+		}
+
+		@Override
+		public void seekToLast() {
+			super.seekToLast();
+			filterOrTransform(super::prev);
+		}
+
+		@Override
+		public void next() {
+			super.next();
+			filterOrTransform(super::next);
+		}
+
+		@Override
+		public void prev() {
+			super.prev();
+			filterOrTransform(super::prev);
+		}
+
+		private void filterOrTransform(Runnable advance) {
+			while (isValid() && (current = stateSnapshotTransformer.filterOrTransform(super.value())) == null) {
```

Comment:
I assume that `super.value()` always returns not `null`? Otherwise this loop could be problematic.

Furthermore, if the iterator became invalid, current is not update and still accessible. I think the use of this iterator always respects to check `valid` first, but we might reset current or throw an exception in the invalid case? Could use some marker `byte[0]` to mark the exception case if you think it makes sense.